executor/src/witgen/machines/block_machine.rs

```diff
@@ -591,18 +644,125 @@ impl<'a, T: FieldElement> WitnessColumnEvaluator<T> for WitnessData<'a, T> {
     }
 }
 
-struct ProcessingSequenceCache {
-    block_size: usize,
-    identities_count: usize,
-    cache: BTreeMap<SequenceCacheKey, Vec<SequenceStep>>,
-}
-
 #[derive(Clone, Debug)]
 struct SequenceStep {
     row_delta: i64,
     identity: IdentityInSequence,
 }
 
+/// Holds the current state of the [DefaultSequenceIterator] and implements the state transition.
+/// For each row delta, we loop over all identities (called a "round") until no more progress
+/// is made.
+struct DefaultSequenceIteratorState {
+    /// Whether this is the first time the iterator is called.
+    is_first: bool,
+    /// Whether any progress was made in the current round.
+    progress_in_current_round: bool,
+    /// The current row delta index.
+    cur_row_delta_index: usize,
+    /// The current identity index.
+    cur_identity_index: usize,
+    /// The number of rounds for the current row delta.
+    /// If this number gets too large, we will assume that we're in an infinite loop and exit.
+    current_round_count: usize,
+}
+
+impl DefaultSequenceIteratorState {
+    fn new() -> Self {
+        DefaultSequenceIteratorState {
+            is_first: true,
+            progress_in_current_round: false,
+            cur_row_delta_index: 0,
+            cur_identity_index: 0,
+            current_round_count: 0,
+        }
+    }
+
+    /// Update the state of the iterator.
+    /// If we're not at the last identity in the current row, just moves to the next.
+    /// Otherwise, starts with identity 0 and moves to the next row if no progress was made.
+    fn update(&mut self, is_last_identity: bool, progress_in_last_step: bool) {
+        if !self.is_first {
+            self.progress_in_current_round |= progress_in_last_step;
+            if is_last_identity {
+                // Start next round
+                if !self.progress_in_current_round || self.current_round_count > 20 {
+                    // Move to next row delta, starting with identity 0.
+                    if self.current_round_count > 20 {
+                        log::warn!("In witness generation for block machine, we have been stuck in the same row for 20 rounds. \
+                                    This is likely a bug in the witness generation algorithm.");
+                    }
+
+                    self.cur_row_delta_index += 1;
+                    self.current_round_count = 0;
+                } else {
+                    // Stay and current row delta, starting with identity 0.
+                    self.current_round_count += 1;
+                }
+                self.cur_identity_index = 0;
+                self.progress_in_current_round = false;
+            } else {
+                // Stay at row delta, move to next identity.
+                self.cur_identity_index += 1;
```

Comment:
I have the feeling that this business of incrementing counters could be better solved with actual iterators.

Reply:
Hmm, I tried:
- Replacing `row_deltas` and `cur_row_delta_index` with a `row_delta_iter: Peekable<IntoIter<i64>>`
- Replacing `cur_identity_index` with a `identities_iter: Peekable<IntoIter<IdentityInSequence>>`

is that what you mean?

I felt that it didn't really get simpler, because we only want to advance the iterators in some cases, so we need to `peek()` a lot. I can give it another try if you still think it would improve clarity.